### ISSUE TITLE
iOS: Users are charged, but we only receive an error

### ISSUE THREAD

**Issue by omsting** [jamesmontemagno/InAppBillingPlugin#176]
We have a problem in iOS with some users that paid for a subscription, but in the app they only receive an error and are not recognized as a subscriber. I think it is linked to what "rich" is explaining here: https://forums.developer.apple.com/thread/6431

## Bug Information

Version Number of Plugin: 2.0.0
Device Tested On: iPhone / iPad
Version of VS: 15.7.4
Version of Xamarin: 4.10

### Steps to reproduce the Behavior
- User tries to subscribe to a product in the app
- There is a problem during the process (e.g. Invalid credit card)
- User solves the problem and complete the purchase

### Expected Behavior
- User goes back to the app and receives a popup indicating that there was an error
- User receives another popup confirming the purchase
- User is now recognized as a subscriber in the app

### Actual Behavior
- User goes back to the app and receives a popup indicating that there was an error
- User is not recognized as a subscriber in the app, but has been charged for the subscription

**Comment by jamesmontemagno** [comment]
You may want to set this: https://github.com/jamesmontemagno/InAppBillingPlugin/blob/master/src/Plugin.InAppBilling.iOS/InAppBillingImplementation.cs#L24 to handle out of band updates.

**Comment by mgoodfellow** [comment]
The solution above does not seem to work, and users that have updated their payment method during the checkout process will still have failed purchases.

**Comment by mgoodfellow** [comment]
Thanks for the reply @jamesmontemagno 

I've been reading that, but as far as I see it, you are already handling the out of band updates then - this PaymentObserver is created and always listening, and you can assign the `OnPurchaseComplete` handler.

This fires when the transaction is successful: https://github.com/jamesmontemagno/InAppBillingPlugin/blob/master/src/Plugin.InAppBilling.iOS/InAppBillingImplementation.cs#L423

It seems to me that we never run the VerifyPurchase stage because we don't have a handler linked for the purchase we aren't expecting: https://github.com/jamesmontemagno/InAppBillingPlugin/blob/master/src/Plugin.InAppBilling.iOS/InAppBillingImplementation.cs#L242

This means our server side bits can never be notified to verify the purchase when this unexpected 2nd transaction comes through?

I'm very much struggling with seeing a solution to this problem at the moment, too many redirected event handlers, and the chain is very hard to follow. Furthermore, Apple don't seem to offer any way to debug this making the process even worse.

If you have any pointers on how I can adapt this library to deal with this issue I would be forever grateful!

**Comment by jamesmontemagno** [comment]
@mgoodfellow I would recommend pulling the source code directly into your project and not use the DLLs to help diagnose the issue that you are having here. You should be able to debug the sandbox approach and see what is happening. I am out of office this week for a conference, but will be back to look at it next week.

**Comment by mgoodfellow** [comment]
@jamesmontemagno Any chance you have had a chance to look at this? We haven't been able to allocate resource (I'm not an expert in iOS platform bits and the only dev for our platform), but would be more than happy to compensate you for your time, while also (hopefully) giving something back to the OSS community for this massively useful library.

**Comment by jamesmontemagno** [comment]
I recently tested this out and really couldn't reproduce it with my sandbox users. Do you have an exact repro that I can walk through and try to debug? 

Something you could do it add a payment observer in your app delegate and monitor all transactions in and out of band yourself. See: https://docs.microsoft.com/en-us/xamarin/ios/platform/in-app-purchasing/

**Comment by mgoodfellow** [comment]
Thanks for the reply @jamesmontemagno - Having spoken to the users who have been affected by this, it is when they change their payment method during the purchase flow (insufficient funds, invalid card etc.). 

I'm not sure of a way to reproduce this on sandbox which is the challenge with this specific issue.

I will look at monitoring all transactions out of band and see if I can capture one of these issues taking place, and log it to our servers. We want to avoid breaking anything on our live deployments obviously which is the challenge with an obscure bug like this!

**Comment by jamesmontemagno** [comment]
So, there are a few things that you can do.

I am hoping that if you set: InAppBillingImplementation.OnPurchaseComplete this should actually get called if something happens out of band and we aren't awaiting on it. I will change this to happen every time a purchase is complete so if you want to do it via event.

The other thing to do is if you get an error when making the purchase you should be able to query your purchases when the page is visible again. It will be in that list if they purchased the item.

Apple requires a restore purchases button so make sure that it gets all the purchases back. That usually solves a lot.

**Comment by mgoodfellow** [comment]
I had added some logging on the `InAppBillingImplementation.OnPurchaseComplete` callback, and it didn't seem to fire. I will use AppCenter to try to capture the event rather than shipping the logs to our servers.

I have put the Restore Purchases functionality in, but nothing ever appears in that list. I assumed this was because the purchase had completed and consumed. (I realise this issue was opened about non-consumables, but in our case it is for a consumable)

**Comment by mgoodfellow** [comment]
Hi @jamesmontemagno were you going to prepare a pre-release package which makes the change to `InAppBillingImplementation.OnPurchaseComplete` so it happens on all events? Sorry, I may have misinterpreted your last message.

**Comment by EmilAlipiev** [comment]
@jamesmontemagno  i want to fork the library. which branch do you suggest? are you now working on twitch version? is it the feature or master yet?

**Comment by jveltd** [comment]
@mgoodfellow are you still seeing issues?  We've had half a dozen failures over the past few days.  I'm reaching out to each user to enquire as to the exact steps taken so I can attempt to repeat the problem.  I'll let you know if we figure out the cause...

**Comment by mgoodfellow** [comment]
@jveltd Yeah, we still have daily issues with this - 95% of our users report they have either:

- Changed payment method during purchase
- Used paypal method during purchase (effectively counts as a payment method change)

Would love to get it fixed, but had no luck as of yet.

**Comment by jveltd** [comment]
@mgoodfellow yep that's pretty much what we're seeing.  I believe all the issues we've had are users who have PayPal as their payment method and there is a a change during the payment process.

We've tested ourselves using PayPal and it was fine so I'm not sure what the users are doing or having to do (update a card? address? something else?) so will continue to ask.

Interestingly, the previous component we were using to handle IAPs didn't seem to suffer this issue.  I might re-visit the old code to see if there are any clues / anything else we can do better.

**Comment by mnxamdev** [comment]
We are having this issue as well. I've noticed it personally when the App Store credit card I was using was expired. I was kicked out of my app and sent to the App Store app to update the payment info. Once I updated the payment info in the App Store app (pressed "Done"), the purchase was executed while still in the App Store but no callback was made to IInAppBillingVerifyPurchase.VerifyPurchase. Additionally, you can get kicked out of the app while purchasing and thrown into the App Store app when Apple updated their ToS and the customer had not previously agreed to the new ToS. Purchase is executed but IInAppBillingVerifyPurchase.VerifyPurchase is not called. Additionally, there seems to be cases where IInAppBillingVerifyPurchase.VerifyPurchase is not called at all even within our app but can't reproduce the issue. This is a serious problem which someone can hopefully fix soon.

**Comment by kjoiner** [comment]
@mnxamdev @jveltd @mgoodfellow @jamesmontemagno  
Hey guys, we ran into this exact issue a few months ago for paypal transactions & when someone had to update their card details during the transaction. Take a look at the code below, TransactionCompleted will be null for out-of-band payments so onPurchaseSuccess never gets called. 

https://github.com/jamesmontemagno/InAppBillingPlugin/blob/1b1876ff92704b37e0cc3a48ecaa8951ffb8142c/src/Plugin.InAppBilling.iOS/InAppBillingImplementation.cs#L446-L453

We pulled in the source locally and changed it to call onPurchaseSuccess when TransactionCompleted is null and it resolved the issue. You will still needed to implement OnPurchaseComplete in the AppDelegate to handle the out-of-band transaction.

``` csharp
case SKPaymentTransactionState.Purchased:
    if (TransactionCompleted != null)
    {
        TransactionCompleted?.Invoke(transaction, true);
    }
    else
    {
        onPurchaseSuccess?.Invoke(transaction.ToIABPurchase());
    }

    SKPaymentQueue.DefaultQueue.FinishTransaction(transaction);
    break;
```

**Comment by jveltd** [comment]
@kjoiner You legend!  Thank you very much for sharing this.  I'll take a look in more detail in the morning when I've got some caffeine flowing through my veins!!

**Comment by mgoodfellow** [comment]
@kjoiner This is amazing - thank you so much for providing a fix!

Can you issue a PR against the repo? In the meantime we will bring the sources in locally and issue some beta builds and monitor

Thanks!

**Comment by mgoodfellow** [comment]
Just thinking about it - to match original behaviour, we shouldn't use the `else` - this could break existing implementations for users if they expect `OnPurchaseSuccess` to always be called.

In fact, we also don't need to check `TransactionCompleted` as this is using null-propagation...

```
case SKPaymentTransactionState.Purchased:
    TransactionCompleted?.Invoke(transaction, true);
    
    onPurchaseSuccess?.Invoke(transaction.ToIABPurchase());

    SKPaymentQueue.DefaultQueue.FinishTransaction(transaction);
    break;
```

**Comment by kjoiner** [comment]
@mgoodfellow @jamesmontemagno Is the expectation here that OnPurchaseComplete is called for both TransactionCompleted and onPurchaseSuccess? Just checking being onPurchaseSuccess was originally only called if TransactionCompleted was not null.

**Comment by mgoodfellow** [comment]
@kjoiner The current implementation was to call `onPurchaseSuccess` if `TransactionCompleted` was not null.

Your implementation changed this behaviour with the `else` statement. This could break existing implementations.

My change makes sure `OnPurchaseSuccess` is invoked regardless.

In your case I would argue the name `OnPurchaseSuccess` is wrong, really it would be `OnOutOfBandPurchaseSuccess`, as it would no longer fire on normal purchase successful events.

Hard to say - but I have chosen to handle the `Verify` step in my `OnPurchaseSuccess` handler in our current test version of the app, rather than after the call to `PurchaseAsync`

**Comment by kjoiner** [comment]
@mgoodfellow I agree with you, I think this can be refactored to remove confusion. If we decide that's the direction we want to go in I can look into it.

**Comment by devology-rob** [comment]
Hi @jamesmontemagno

I'm trying to understand what we need to do to support out-of-band iOS messages for non-consumable and consumable purchases (we are not concerned with Subscriptions at this time).  There appears to be two potential solutions to this issue.

**Solution 1** Pull request #204 'Always calls OnPurchaseSuccess if defined for out-of-band purchases' - does this end up calling OnPurchaseComplete as part of the lifecycle / what's the relationship to solution 2?

**Solution 2** The original advice was to add an iOS specific OnPurchaseComplete listener to handle out of band updates.

So should we only update our InAppBillingPlugin nugets to benefit from solution 1 (PR #204) and then out-of-bound will 'just work' and/or do we need to register an OnPurchaseComplete listener?

### Solution 2 - OnPurchaseComplete listener

Will the OnPurchaseListener fire an event for normal purchases and out-of-band purchases?  For example we already have business logic that handles the purchasing...

```
	// #A
	public Task<bool> VerifyPurchase(string signedData, string signature, string productId = null, string transactionId = null)
	{
	    PurchaseManager.SignedData = signedData;
		
		#if __ANDROID__

	        // On Android we can verify the signed data before passing to the server
	        return Task.FromResult(Plugin.InAppBilling.InAppBillingImplementation.InAppBillingSecurity.VerifyPurchase(
	            getGoogleApiKey(), 
	            signedData, 
	            signature));
		#else
	    	return Task.FromResult(true); // Let the server verify this
		#endif        
	}

	...

	// #B
	var purchase = inAppBilling.PurchaseAsync(productIdAkaSKU, ItemType.InAppPurchase, DeveloperPayload, PurchasedManager);

	// #C
 	if (purchase.State == PurchaseState.Purchased)
    {
        // Verify token/signed data captured from verification step on the server
        // and receive the payload if successful
        // (passes the stored signedData to the server captured during verification)

        // Tell user success
    } else {
    	// Tell user fail
    }

```

So should we move the logic in #C into OnPurchaseComplete exclusively, or is this only for the out-of-band ones? 

Does OnPurchaseListener include everything inAppBilling.PurchaseAsync and VerifyPurchase uses? 

- Signed Data
- Signature
- Transaction ID

What about determining the

- Product ID (SKU)
- Developer payload

I note that @mgoodfellow said he still had issues after adding a OnPurchaseComplete listener (22nd August), but his issue was related to subscriptions, which we are not using.   However his code seemed to do some IoC lookup of the IInAppBillingVerifyPurchase which our equivallent is only section #A in our code.   But this on it's own doesn't 'do' anything - so I suspect we need to invoke our usual code that completes the purchase (#C) within OnPurchaseComplete and maybe that's the step that @mgoodfellow might have missed? 

### Verifying in general

We verify the receipt on the server (#C) based on the verification token/signed data captured in the IInAppBillingVerifyPurchase::VerifyPurchase method (#A), that's reached when we successfully invoke the purchase (#B).  Therefore we don't verify the iOS token in the app per say, so...

 - Premium app upgrade (non consumable) - we just unlock premium if purchase.State == PurchaseState.Purchased.  We are aware some jailbroken devices could get the premium upgrade for free

 - For consumable products (that actually cost us money) we call through to our server passing the signedData through, the server then verifies the SKU, invoice dates, signed data, etc and if successful we will make a back-end purchase and then return back the data payload with the purchased data, or an error message if verification fails.

Presumably this strategy is okay? Is there a copy&paste equivalent of the Android verification step that we can do for iOS to perform verification in the app too?

### How to invoke an out-of-band purchase?

Whilst our customers have been exhibiting these issues, we can't find a way of initiating them to aid our testing, do you have any recommendations?

### Exception handling

You have also mentioned exception handling around iOS calls somewhere in the issues (but I can't find that reference right now), but I wasn't sure on the context of this.  Was the recommendation of using try/catch blocks in context to OnPurchaseComplete, IInAppBillingVerifyPurchase implementation, or surrounding all invocations of the InAppBillingPlugin?

**Comment by mgoodfellow** [comment]
> I note that @mgoodfellow said he still had issues after adding a OnPurchaseComplete listener (22nd August), but his issue was related to subscriptions, which we are not using.

@devology-rob Actually, I hijacked this thread - I'm using consumables! 

> I have put the Restore Purchases functionality in, but nothing ever appears in that list. I assumed this was because the purchase had completed and consumed. (I realise this issue was opened about non-consumables, but in our case it is for a consumable - I honestly thought I had mentioned that, but it appears I never did! Apologies for this!!) 

I now do all verification via the OnPurchaseComplete listener for iOS, and for Android it is done after the call to `CreatePurchaseAsync` (sic)

**Comment by devology-rob** [comment]
Hey thanks @mgoodfellow.  It sounds like for Android we could leave our code exactly as-is (we have one `PurchaseManager` shared across both platforms for our purchase business logic) but for iOS we could ignore the result after awaiting the `inAppBilling.PurchaseAsync` and rely exclusively on the OnPurchaseComplete listener instead?  

I'm not sure how `CreatePurchaseAsync` relates to `PurchaseAsync`?

I'd like to test the out-of-band workflow, but alas I can't find a way of doing that - have you found a way of performing a purchase workflow out-of-band yet?

**Comment by mgoodfellow** [comment]
Sorry, it is `PurchaseAsync` I'm working from memory - sorry for the confusion!

> It sounds like for Android we could leave our code exactly as-is (we have one PurchaseManager shared across both platforms for our purchase business logic) but for iOS we could ignore the result after awaiting the inAppBilling.PurchaseAsync and rely exclusively on the OnPurchaseComplete listener instead?

This is almost exactly what we have and do

> I'd like to test the out-of-band workflow, but alas I can't find a way of doing that - have you found a way of performing a purchase workflow out-of-band yet?

Same, but no I haven't managed - some of my customers tell me it happens with PayPal - but I don't know how to make iTunes accept PayPal to try this out myself

**Comment by devology-rob** [comment]
> Sorry, it is `PurchaseAsync` I'm working from memory - sorry for the confusion!

😅 I thought I was missing something.
 
> This is almost exactly what we have and do

That's re-assuring, thanks.
 
> Same, but no I haven't managed - some of my customers tell me it happens with PayPal - but I don't know how to make iTunes accept PayPal to try this out myself

- Go to Settings
- Click on your name (Apple ID, iCloud, iTunes & App Store)
- Payment & Shipping
- Click on Payment Method
- Scroll to the bottom and click Change Payment Method
- Choose PayPal
- Follow all the steps

Once completed I would advise you to go back into that screen and verify it's set - I've done the full process once and it acted as though it was changed, but it wasn't 🙄

The issue we had once we set PayPal is that it still 'just worked'.  I don't know whether there's some sort of two-factor authentication some users have that change the workflow, but to-date we haven't sussed it out.

**Comment by mgoodfellow** [comment]
@mnxamdev You can pull the nuget generated via appveyor directly into your project for now - I got it from here: https://ci.appveyor.com/project/JamesMontemagno/inappbillingplugin/builds/21738621

I'm sure a new nuget will be published soon, but that is the fastest way to get going - I then extracted out the libs and imported into as references into the various projects! Slightly painful - but a lot easier than bringing in the repo ;) - if you have a private nuget feed, you can always push it onto there as well

Yes, I think the docs should be updated - with this patch, you can carry on as before not using OnPurchaseComplete handler, but you won't get the benefits of picking up out-of-band purchases.

If you _do_ use the OnPurchaseComplete it might be worth only handling verification / processing on the callback from this handler, and not after your call to `PurchaseAsync`

For a rough example, this is what my callback looks like:

```csharp
InAppBillingImplementation.OnPurchaseComplete = async purchase =>
            {
                try
                {
                    var verify = Mvx.Resolve<IInAppBillingVerifyPurchase>();

                    // Get the receipt data for (server-side) validation.
                    // See: https://developer.apple.com/library/content/releasenotes/General/ValidateAppStoreReceipt/Introduction.html#//apple_ref/doc/uid/TP40010573
                    var receiptUrl = NSData.FromUrl(NSBundle.MainBundle.AppStoreReceiptUrl);
                    var receipt = receiptUrl.GetBase64EncodedString(NSDataBase64EncodingOptions.None);

                    var verificationResult = await verify.VerifyPurchase(
                        receipt,
                        string.Empty,
                        purchase.ProductId,
                        purchase.Id);
                }
                catch (Exception ex)
                {
                    //
                }
            };
```

I set this up in the `AppDelegate.cs` file - if verification passes, in my app I fire off a message to handle updates, and display a message to the user.

Android purchase flow verifies after call to `PurchaseAsync` so it is a little bit confusing as there are 2 different flows.

**Comment by jamesmontemagno** [comment]
@mgoodfellow can you add some documentation as to how you did this :)

**Comment by devology-rob** [comment]
Hi @jamesmontemagno and @mgoodfellow I usually add nugets straight into my project and just 'update' them when a stable update is available, so I'm not comfortable with the process that would be required to 'pull the repo' into our projects - could someone please elaborate?

FYI I did try building your solution then copying over the generate dlls into my project.  For whatever reason that didn't work so I tried adding the project instead and referenced that instead of the nuget, but that didn't work either.  I'm not sure whether either of those techniques are what you refer to when 'pull the repo' is mentioned?

As for appveyor - was this build triggered from your own clone of James' repo?  I'm just curious how this fits in.  Also are the steps to pull in your nuget different to pulling in the official one?

@jamesmontemagno Is there an ETA for updating the official nuget?

**Comment by mnxamdev** [comment]
@mgoodfellow thanks so much for your contribution and explanation. I looked and the nuget seems to be appended as "beta". I have concerns given that the code here is merged in master but not fully released to official nuget. Is there a QA process that happens @jamesmontemagno as part of this? I'm aware this is open source, asking for a friend. Seriously though, this plugin affects revenue so I'm asking.

**Comment by kjoiner** [comment]
@devology-rob Pull the repo typically refers to cloning the repo and adding the projects to your solution locally. You'll want to remove any references to the nuget packages to avoid conflicts. I have the in app billing projects in a solution folder to keep them organized. Once you bring them in you'll add references to them from your existing projects. As long as you don't change the namespaces everything should work. This is how I stepped through the in app billing implementation and found the original bug. You mentioned that you tried this but it didn't work. What error did you get?

**Comment by devology-rob** [comment]
Thanks @kjoiner - I'll retry today.

>  You mentioned that you tried this but it didn't work. What error did you get?

If I recall it was the fact I couldn't add the event listener in the iOS project because 'OnPurchaseComplete' was not offered as a intellisense-completion and it wouldn't compile (InAppBillingImplementation.OnPurchaseComplete). I added using statements and in the end I assumed I was doing something wrong, but I will now retry with confidence that this was the way that you achieved it.

**Comment by devology-rob** [comment]
Hi @kjoiner 

Oh here we go again 🤪, I get told that "The name 'InAppBillingImplementation' does not exist in the current context", yet Visual Studio shows me the intellisence bubble that shows clearly the types returned and events that I can add to...

<img width="1229" alt="screenshot 2019-02-07 at 09 46 37" src="https://user-images.githubusercontent.com/18516600/52402891-4b2df780-2abd-11e9-866e-58a9c3a50e94.png">

Within the iOS project I pull in the three Plugin.InAppBilling dependencies...

<img width="244" alt="screenshot 2019-02-07 at 09 44 18" src="https://user-images.githubusercontent.com/18516600/52402789-0dc96a00-2abd-11e9-8644-34a1ca1fad5a.png">

The bit I didn't quite follow is that in my original NuGet based projects I have a common PCL used between Android and iOS and in there was only one Nuget dependency, Plugin.InAppBilling, but to get this compiling by using the local GitHub cloned projects I needed to reference Plugin.InAppBilling and Plugin.InAppBilling.Abstractions.

In Android and iOS I kept adding project dependencies until compilation worked, never quite sure why I needed to add more that the top-level plugin name originally present.

If I look at the referenced Plugin.InAppBilling.iOS library in the iOS project I can clearly see the class and event as expected 🤷🏻‍♂️

<img width="955" alt="screenshot 2019-02-07 at 09 56 04" src="https://user-images.githubusercontent.com/18516600/52403432-9d234d00-2abe-11e9-86fc-0cc93d6cf12a.png">

Any pointers about which projects to add at common PCL, Android and iOS projects will be appreciated - at the moment it feels like I need to do some clean+rebuild, quit visual studio, delete obj folders, throw salt over my shoulder etc... Not sure why the error wasn't there and now it has come back.

**Comment by devology-rob** [comment]
And as soon as I sent that I wondered why I needed Plugin.InAppBilling, so removed it and now it compiles - guidance still appreciated 😂

**Comment by AdelaPetrut** [comment]
I also am experiencing the same problem.  Does the newly created 2.1.0.187-beta contain the correction? When can we estimate an official release containing this?
Thank you very much everyone!

**Comment by jamesmontemagno** [comment]
Yes, it always has this fix.

**Comment by AdelaPetrut** [comment]
Thank you @jamesmontemagno for the response. 
Unfortunately , I included the beta version, created a new Store build (in order to be available to test) but the error is still happening:
The error type is PurchaseError.UserCancelled -> "The user cancelled the purchase"

The steps to reproduce are the following:
1. Try to purchase (without a card associated to the account)
2. You are directed to the Account Setting section to include the card credentials
3.  You receive the message "Your purchase was successful"
4. Inside app afterwards. from the InAppBilling .dll we receive "The user cancelled the purchase" error.

Only I i press again the purchase (message "you are currently already subscribed" will appear -> then it works without error from the .dll)

**Comment by AdelaPetrut** [comment]
I may have missed a very important information mentioned by @mgoodfellow 

To include in the AppDelegate:
InAppBillingImplementation.OnPurchaseComplete = async purchase =>
            {
                try
                {
                    var verify = Mvx.Resolve<IInAppBillingVerifyPurchase>();

                    // Get the receipt data for (server-side) validation.
                    // See: https://developer.apple.com/library/content/releasenotes/General/ValidateAppStoreReceipt/Introduction.html#//apple_ref/doc/uid/TP40010573
                    var receiptUrl = NSData.FromUrl(NSBundle.MainBundle.AppStoreReceiptUrl);
                    var receipt = receiptUrl.GetBase64EncodedString(NSDataBase64EncodingOptions.None);

                    var verificationResult = await verify.VerifyPurchase(
                        receipt,
                        string.Empty,
                        purchase.ProductId,
                        purchase.Id);
                }
                catch (Exception ex)
                {
                    //
                }
            };

I will try and come up with details

**Comment by AdelaPetrut** [comment]
Thank you all for the details and help.

I managed to successfully resolve the problem
Steps included:
0. Upgrade to the latest Beta version
1. Include in the AppDelegate class the callback to InAppBillingImplementation.OnPurchaseComplete
2. And, as @mgoodfellow mentioned: 
Only handling verification / processing on the callback from this handler, and not after your call to `PurchaseAsync` - otherwise you will be handling duplicates in most cases!

**Comment by mnxamdev** [comment]
Hi,
Sorry to be commenting on a closed issue but we are now using the Beta 3 nuget package and just getting back to Android on this. Are we still needing to take different paths on iOS vs. Android? We are using the Plugin.InAppBilling.InAppBillingImplementation.OnPurchaseComplete handler on iOS to then verify purchases, so we don't pass in IInAppBillingVerifyPurchase into the CrossInAppBilling.Current.PurchaseAsync method. Is there an Android equivalent to this in Beta 3? If not, then do we pass in IInAppBillingVerifyPurchase into PurchaseAsync on Android or is there another method we need to go about for verifying purchases?
Thanks in advance.

> @mnxamdev You can pull the nuget generated via appveyor directly into your project for now - I got it from here: https://ci.appveyor.com/project/JamesMontemagno/inappbillingplugin/builds/21738621
> 
> I'm sure a new nuget will be published soon, but that is the fastest way to get going - I then extracted out the libs and imported into as references into the various projects! Slightly painful - but a lot easier than bringing in the repo ;) - if you have a private nuget feed, you can always push it onto there as well
> 
> Yes, I think the docs should be updated - with this patch, you can carry on as before not using OnPurchaseComplete handler, but you won't get the benefits of picking up out-of-band purchases.
> 
> If you _do_ use the OnPurchaseComplete it might be worth only handling verification / processing on the callback from this handler, and not after your call to `PurchaseAsync` - otherwise you will be handling duplicates in most cases!
> 
> For a rough example, this is what my callback looks like:
> 
> ```cs
> InAppBillingImplementation.OnPurchaseComplete = async purchase =>
>             {
>                 try
>                 {
>                     var verify = Mvx.Resolve<IInAppBillingVerifyPurchase>();
> 
>                     // Get the receipt data for (server-side) validation.
>                     // See: https://developer.apple.com/library/content/releasenotes/General/ValidateAppStoreReceipt/Introduction.html#//apple_ref/doc/uid/TP40010573
>                     var receiptUrl = NSData.FromUrl(NSBundle.MainBundle.AppStoreReceiptUrl);
>                     var receipt = receiptUrl.GetBase64EncodedString(NSDataBase64EncodingOptions.None);
> 
>                     var verificationResult = await verify.VerifyPurchase(
>                         receipt,
>                         string.Empty,
>                         purchase.ProductId,
>                         purchase.Id);
>                 }
>                 catch (Exception ex)
>                 {
>                     //
>                 }
>             };
> ```
> 
> I set this up in the `AppDelegate.cs` file - if verification passes, in my app I fire off a message to handle updates, and display a message to the user.
> 
> Android purchase flow verifies after call to `PurchaseAsync` so it is a little bit confusing as there are 2 different flows.

**Comment by alirazamsp** [comment]
@jamesmontemagno  @mgoodfellow @kjoiner @EmilAlipiev @jveltd 
I am facing this same issue in Xamarin Forms app where users report that they have bought item, they have been charged but they did not get access to the product/service. I am using InAppBilling stable version 2.0.0 in my app. AddPaymentToDB() function adds payment record to database which is not fired in this case.  Below is the code which I am using: 

` public async Task Purchase()
        {
            try
            {
                    var productId = GetCourseProductId(courseDetails.price);
                    var connected = await CrossInAppBilling.Current.ConnectAsync();
                    if (!connected)
                    {
                        //Couldn't connect to billing, could be offline, alert user
                        Settings.Message = "Could not connect to billing right now, it could be offline. Try again later.";
                        return;

                    }

                    //try to purchase item
                    var purchase = await CrossInAppBilling.Current.PurchaseAsync(productId, ItemType.InAppPurchase, "apppayload");
                    if (purchase == null)
                    {
                       await home.DisplayAlert("Error", "Could not complete your purchase. Please try again.","Ok");
                       return;
                    }
                    else
                    {
                        //Purchased, save this information
                        var id = purchase.Id;
                        string serializedObject = JsonConvert.SerializeObject(purchase);
                    // send an API call to save record to DB

                    var result = AddPaymentToDB(id, serializedObject).GetAwaiter().GetResult();
                    await home.DisplayAlert("Success","You have successfully purchased this course. Your payment ID is :" + purchase.Id,"Ok");
                    await ShowCourse(result.data);
                }
                //}
            
            }
            catch (InAppBillingPurchaseException ex)
            {
                switch (ex.PurchaseError)
                {
                    case PurchaseError.AppStoreUnavailable:
                        HideShowIndicator(false);
                        await home.DisplayAlert("Error","Currently the app store seems to be unavailble. Try again later.","Ok");
                        break;
                    case PurchaseError.BillingUnavailable:
                        HideShowIndicator(false);
                        await home.DisplayAlert("Error", "Billing seems to be unavailable, please try again later.", "Ok");
                        break;
                    case PurchaseError.PaymentInvalid:
                        HideShowIndicator(false);
                        await home.DisplayAlert("Error", "Payment seems to be invalid, please try again.", "Ok");
                        break;
                    case PurchaseError.PaymentNotAllowed:
                        HideShowIndicator(false);
                        await home.DisplayAlert("Error", "Payment does not seem to be enabled/allowed, please try again.", "Ok");
                        break;

                    case PurchaseError.UserCancelled:
                        HideShowIndicator(false);
                        break;

                    default:
                        HideShowIndicator(false);
                        await home.DisplayAlert("Error", ex.Message, "Ok");
                        break;
                }
                isProcessingPaymentRequest = false;

            }
            catch(Exception ex)
            {
                Settings.Message = ex.Message;
                isProcessingPaymentRequest = false;
            }
            finally
            {
                await CrossInAppBilling.Current.DisconnectAsync();
            }
        }`


Has this issue been fixed in latest beta release of this plugin or do i need to do anything additional in order to resolve this problem ?  Thanks.

**Comment by jamesmontemagno** [comment]
You can check the error code for already purchased maybe. Or loop through existing purchases and see if it is there.

**Comment by alirazamsp** [comment]
This is a consumable item so I believe it will be purchased again upon next purchase. I asked my customers, this issue happened when they had to add credit card with their Apple ID or their card was expired and they had to change their credit card for Apple ID. Does Beta pre-release 3.0.0 has this issue fixed in it ? 

**You can check the error code for already purchased maybe. Or loop through existing purchases and see if it is there.**
Did you mean I should make a send a call for existing purchases for current Apple Id and then check for errors if any. Then, I should go for purchase of the item ? 

Do I need to add InAppBillingImplementation.OnPurchaseComplete in AppDelegate ? 
Can you please confirm if there is not anything wrong in the code above?
